src/flashcart/ed64/ed64.c

```diff
@@ -0,0 +1,313 @@
+
+    size_t rom_size = f_size(&fil);
+
+    // FIXME: if the cart is not V3 or X5 or X7, we need probably need to - 128KiB for save compatibility.
+    // Or somehow warn that certain ROM's will have corruption due to the address space being used for saves.
+    // Conker's Bad Fur Day doesn't have this issue because eeprom data is at a fixed address in pif ram.
+    if (rom_size > MiB(64)) {
+        f_close(&fil);
+        return FLASHCART_ERR_LOAD;
+    }
+
+    if (rom_size == MiB(64)) {
+        ed64_save_type_t type = ed64_ll_get_save_type();
+        switch (type) {
+            case SAVE_TYPE_SRAM:
+                rom_size -= KiB(32);
+                break;
+            case SAVE_TYPE_SRAM_128K:
+            case SAVE_TYPE_FLASHRAM:
+                rom_size -= KiB(128);
+                break;
+            case SAVE_TYPE_DD64_CART_PORT:
+                break;
+            default:
+                break;
+        }
+    }
+    
+    size_t sdram_size = rom_size;
+
+    size_t chunk_size = MiB(1);
+    for (int offset = 0; offset < sdram_size; offset += chunk_size)
+    {
+        size_t block_size = MIN(sdram_size - offset, chunk_size);
+        if (f_read(&fil, (void *)(ROM_ADDRESS + offset), block_size, &br) != FR_OK) {
+            f_close(&fil);
+            return FLASHCART_ERR_LOAD;
+        }
+        if (progress) {
+            progress(f_tell(&fil) / (float)(f_size(&fil)));
+        }
+    }
+
+    if (f_close(&fil) != FR_OK) {
+        return FLASHCART_ERR_LOAD;
+    }
+
+    return FLASHCART_OK;
+}
+
+static flashcart_err_t ed64_load_file (char *file_path, uint32_t rom_offset, uint32_t file_offset)
+{
+    FIL fil;
+    UINT br;
+
+    if (f_open(&fil, strip_sd_prefix(file_path), FA_READ) != FR_OK) {
+        return FLASHCART_ERR_LOAD;
+    }
+
+    fix_file_size(&fil);
+
+    size_t file_size = f_size(&fil) - file_offset;
+
+    // FIXME: if the cart is not V3 or X5 or X7, we need probably need to - 128KiB for save compatibility.
+    // Or somehow warn that certain ROM's will have corruption due to the address space being used for saves.
+
+    if (file_size > (MiB(64) - rom_offset)) {
+        f_close(&fil);
+        return FLASHCART_ERR_ARGS;
+    }
+
+    if (f_lseek(&fil, file_offset) != FR_OK) {
+        f_close(&fil);
+        return FLASHCART_ERR_LOAD;
+    }
+
+    if (f_read(&fil, (void *)(ROM_ADDRESS + rom_offset), file_size, &br) != FR_OK) {
+        f_close(&fil);
+        return FLASHCART_ERR_LOAD;
+    }
+    if (br != file_size) {
+        f_close(&fil);
+        return FLASHCART_ERR_LOAD;
+    }
+
+    if (f_close(&fil) != FR_OK) {
+        return FLASHCART_ERR_LOAD;
+    }
+
+    return FLASHCART_OK;
+}
+
+static flashcart_err_t ed64_load_save (char *save_path) {
+    
+    FIL fil;
+    UINT br;
+
+    if (f_open(&fil, strip_sd_prefix(save_path), FA_READ) != FR_OK) {
+        f_close(&fil);
+        return FLASHCART_ERR_LOAD;
+    }
+
+    size_t save_size = file_get_size(strip_sd_prefix(save_path));
```

Comment:
No need to add `strip_sd_prefix` here. `file_*` functions in `utils/fs.c` already do that. `strip_sd_prefix` should be used only where function expects path to not include `sd:/` prefix, and that generally applies only to the FatFs functions.

Reply:
By the way, if you opened file with `f_open` you already have file size accessible with `f_size(&fil);`. See SC64 implementation: https://github.com/Polprzewodnikowy/N64FlashcartMenu/blob/553e986a380b9ed1cb81253c644ac42ed3ec17b2/src/flashcart/sc64/sc64.c#L418